File: data-structure/linked_list/cyclic.js

const isCyclic = list => {
    let slow = list._head;
    let fast = list._head.next;

    while (true) {
        if (!fast || !fast.next) {
            return false;
        }

        if (slow !== fast) {
            slow = slow.next;
            fast = fast.next.next;
        } else {
            return true;
        }
    }
};

// 17 -> 3 -> 100 -> 42 -> 5 -> 1972

list.addNode(1972);
list.addNode(5);
list.addNode(42);
const n100 = list.addNode(100);
list.addNode(3);
list.addNode(17);
const n666 = list.appendNode(666);

list.insertAfter(n100, 200);
// list.print();

// 17 -> 3 -> 100 -> 200 -> 42 -> 5 -> 1972 -> 666

list.printNodes();

console.log(isCyclic(list));
n666.next = n100;
console.log(isCyclic(list));

File: c/data_structures/linked_list/cyclic.c
#include "../../include/linkedList.c"

int isCyclic(struct list *list) {
    struct node *slow = list->head;
    struct node *fast = list->head->next;

    while (1) {
        if (!fast || !fast->next)
            return 0;

        if (fast == slow || fast->next == slow)
            return 1;

        slow = slow->next;
        fast = fast->next->next;
    }
}

int main(void) {
    struct list *list = makeList();
    struct node *node;

    addNode(&list, 0);
    addNode(&list, 1);
    addNode(&list, 2);
    addNode(&list, 3);
    addNode(&list, 4);
    addNode(&list, 5);
    addNode(&list, 6);
    node = addNode(&list, 7);
    addNode(&list, 8);
    addNode(&list, 9);
    addNode(&list, 10);
    addNode(&list, 11);
    addNode(&list, 12);
    addNode(&list, 13);
    addNode(&list, 14);

    // Create a cyclic list.
    list->tail->next = node;

    if (isCyclic(list))
        printf("is cyclic");
    else
        printf("is not cyclic");

    return 0;
}

